In [1]:
import numpy as np

## Testing Markov Model by converting costs to normalised Probability

In [2]:
data = np.array([[1,0.5,1.45,0.75],[1.95,1,3.1,1.49],[0.67,0.31,1,0.48],[1.34,0.64,1.98,1]])

In [3]:
data

array([[1.  , 0.5 , 1.45, 0.75],
       [1.95, 1.  , 3.1 , 1.49],
       [0.67, 0.31, 1.  , 0.48],
       [1.34, 0.64, 1.98, 1.  ]])

In [4]:
from sklearn.preprocessing import normalize

In [5]:
data1=normalize(data,norm='l1',axis=0)

In [6]:
data1

array([[0.2016129 , 0.20408163, 0.19256308, 0.2016129 ],
       [0.39314516, 0.40816327, 0.41168659, 0.40053763],
       [0.13508065, 0.12653061, 0.13280212, 0.12903226],
       [0.27016129, 0.26122449, 0.26294821, 0.2688172 ]])

In [7]:
initial=np.array([0,0,0,1])

In [8]:
t1=np.dot(initial,data1)
np.dot(initial,data1)

array([0.27016129, 0.26122449, 0.26294821, 0.2688172 ])

In [9]:
t2=np.dot(t1,data1)
np.dot(t1,data1)

array([0.26531036, 0.26524983, 0.26517079, 0.26528973])

In [10]:
t3=np.dot(t2,data1)
np.dot(t2,data1)

array([0.26526214, 0.26526261, 0.26526148, 0.26526256])

In [11]:
t4=np.dot(t3,data1)
np.dot(t3,data1)
#Results were not as good as below coded logic.

array([0.26526235, 0.26526236, 0.26526236, 0.26526236])

## Accepted Logic on States.

In [12]:
#Accepted Logic
trades = [[1, 0.5, 1.45, 0.75],
          [1.95, 1, 3.1, 1.49],
          [0.67, 0.31, 1, 0.48],
          [1.34, 0.64, 1.98, 1]
          ]

num_trades = 5
curr_prod = 3
possible_trades = set()
def get_best_trade(curr_prod, past_trades):

    if len(past_trades) == num_trades:
        past_trades[0] *= trades[curr_prod][3]
        past_trades.append(3)
        #print(past_trades)
        possible_trades.add(tuple(past_trades))
        return

    for trade_prod, value in enumerate(trades[curr_prod]):
        #print(trade_prod,value)
        past_trades[0] *= value
        get_best_trade(trade_prod, past_trades + [trade_prod])
        past_trades[0] /= value


if __name__=="__main__":
    get_best_trade(curr_prod, [1])
    best_trade = max(possible_trades, key=lambda x: x[0])
    print(best_trade)



(1.0436925000000001, 0, 1, 2, 0, 3)


##### Achieved 1.04% additional profit on cash at the end.
##### Problem demands to fix initial and final state as Pearl (here currency) with maximum trade 5.